flang/include/flang/Tools/CLOptions.inc

```diff
@@ -304,9 +330,7 @@ inline void createDebugPasses(
 inline void createDefaultFIRCodeGenPassPipeline(
     mlir::PassManager &pm, MLIRToLLVMPassPipelineConfig config) {
   fir::addBoxedProcedurePass(pm);
-  pm.addNestedPass<mlir::func::FuncOp>(
-      fir::createAbstractResultOnFuncOptPass());
-  pm.addNestedPass<fir::GlobalOp>(fir::createAbstractResultOnGlobalOptPass());
+  addNestedPassToAllTopLevelOperations(pm, fir::createAbstractResultOptPass);
```

Comment:
Thanks for review.

I think the idea is that `addNested<OP>` could work on a subset of the operations supported by the pass (reported by `canBeScheduled`). Conceptually, `addNested<OP>` means you want to run the pass on this operation in this particular pipeline. But a different pipeline could be constructed which tries to use the pass on a different operation type (e.g. `fir-opt --abstract-result module.mlir`). `canScheduleOn` guards against running pipelines on operation types which the pass is not intended for.

We have to implement `canScheduleOn` because it is pure virtual in `mlir::Pass`. Even if there were a way to tell the pass manager to run this pass on every operation on which the pass is supported, this would have to be implemented by calling `canScheduleOn` with every operation and then only scheduling the pass on supported operations. Unfortunately, `canScheduleOn` is implemented with string comparisons (it is defined as always taking a `RegisteredOperationName` argument) so I would prefer the bit of duplication so that we can limit these string comparisons.

Reply:
> canScheduleOn guards against running pipelines on operation types which the pass is not intended for.

In that case, I think it should be runnable on any op (the ReturnOp handling prevents that currently since you noticed it does not work on ModuleOp).

> We have to implement canScheduleOn because it is pure virtual in mlir::Pass.

You do not need to in that case because AbstractResultOptPass actually inherits from `mlir::OperationPass<>` that defines canScheduleOn [here](https://github.com/llvm/llvm-project/blob/74e07ab523122d6a8347b25770062ab331b6bb84/mlir/include/mlir/Pass/Pass.h#L373) in a way that it would say true to any operation it is being schedule on. The .td `Pass<>` syntax actually creates a pass inheriting from `mlir::OperationPass<>`, not directly from `mlir::Pass` (see [here](https://github.com/llvm/llvm-project/blob/98427264cab558d285c7222f1f25b904124c6b93/mlir/include/mlir/Pass/PassBase.td#L94C3-L94C38)).

So all in all, I am OK with your patch, it is an improvement, and it could be further improved by modifying the `ReturnOp` handling and removing the canScheduleOn restriction.

Reply:
Ahh I had missunderstood the tablegen change I made. I thought it made it inherit mlir::Pass, but you are right it is `mlir::OperationPass<>`.

Thanks for explaining. I will see if I can create a nested pass pipeline inside the AbstractResultPass so that even when run on a module it behaves the same way as if you ran the old function and global passes. And I agree that the `canScheduleOn` should be removed if possible.